promptsource/templates.py

```diff
@@ -46,6 +46,179 @@ def most_frequent(items):
+        rtemplate = env.from_string(jinja)
+
+        # Replaces any occurrences of the "|||" separator in the example, which
+        # which will be replaced back after splitting
+        pipe_protector = "3ed2dface8203c4c9dfb1a5dc58e41e0"
+        protected_example = {
+            key: value.replace("|||", pipe_protector) if isinstance(value, str) else value
+            for key, value in example.items()
+        }
+
+        # Adds in answer_choices variable
+        if "answer_choices" in protected_example:
+            raise ValueError("Example contains the restricted key 'answer_choices'.")
+        protected_example["answer_choices"] = self.answer_choices
+
+        # Renders the Jinja template
+        rendered_example = rtemplate.render(**protected_example)
+
+        # Splits on the separator, and then replaces back any occurrences of the
+        # separator in the original example
+        return [part.replace(pipe_protector, "|||") for part in rendered_example.split("|||")]
+
+    class Metadata(yaml.YAMLObject):
+        """
+        Metadata for a prompt template.
+        """
+
+        yaml_tag = "!TemplateMetadata"
+
+        def __init__(
+            self,
+            task_format: str,  # dropdown choice of 'Classification', 'Generation', 'Extraction'
+            task_template: bool,
+            metric: Optional[List[str]] = None,
+            # if task_format == classification
+            choices_in_prompt: Optional[bool] = None,
+            # Internal flags
+            _do_train: Optional[bool] = None,
+            _do_eval: Optional[bool] = None,
+        ):
+            """
+            Initializes template metadata.
+
+            In the following, trivial choices are defined as Yes/No, True/False,
+            etc. and nontrivial choices are other types of choices denoted in
+            the answer_choices field.
+
+            :param task_format: Either "Classification" (output is from a set of labels),
+                "Generation" (output is open ended and not contained in the prompt),
+                or "Extraction" (output is open ended but contained in the prompt,
+                including indexed choices)
+            :param task_template: If True, this prompt asks a model to perform the original task designed for
+                this dataset.
+            :param metric: List of strings denoting metrics to use for evaluation
+            :param choices_in_prompt: If True, the answer choices are included in the templates such that models
+                see those choices in the input. Only applicable to classification tasks.
+            :param _do_train: internal parameter. If True, this template should be
+                used for training
+            :param _do_train: internal parameter. If True, this template should be
+                used for evaluation
+            """
+            self.task_format = task_format
+            self.metric = metric
```

Comment:
are we updating this anywhere in the cleaning pass?

Reply:
Currently, no. @awebson and I were thinking it would be a placeholder that we could update as we do more evaluations. We'd need to establish some format for recording the metric(s). Also might require more paper reading to know for sure, so would be slower annotation. Could include if people would find it helpful.